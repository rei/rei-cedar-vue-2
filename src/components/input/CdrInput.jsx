import clsx from 'clsx';
import size from '../../mixins/size';
import propValidator from '../../utils/propValidator';
import CdrLabelStandalone from '../labelStandalone/CdrLabelStandalone';
import CdrFormError from '../formError/CdrFormError';
import style from './styles/CdrInput.scss';
/**
 * Cedar 2 component for input
 * **NOTE:** `v-model` is required.
 * @version 1.0.0
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrInput',
  components: {
    CdrLabelStandalone,
    CdrFormError,
  },
  mixins: [size],
  inheritAttrs: false,
  props: {
    /**
     * `id` for the input that is mapped to the label `for` attribute.
     *  If one is not provided, it will be auto generated.
    */
    id: String,
    /**
     *  'type' attribute for the input as defined by w3c.
     *  Only supporting text|email|number|password|search|url.
     *  The increment/decrement webkit psuedo element is hidden for number.
    */
    type: {
      type: [String],
      default: 'text',
      validator: (value) => propValidator(
        value,
        ['text', 'email', 'number', 'password', 'search', 'url', 'tel'],
      ),
    },
    /**
     * Label text. This is required for a11y even if hiding the label with `hideLabel`.
    */
    label: {
      type: String,
      required: true,
    },
    errorRole: {
      type: String,
      required: false,
      default: "status",
    },
    /**
     * Removes the label element but sets the input `aria-label` to `label` text for a11y.
    */
    hideLabel: Boolean,
    /**
     * Number of rows for input.  Converts component to text-area if rows greater than 1.
    */
    rows: Number,
    // Set which background type the input renders on
    background: {
      type: [String],
      default: 'primary',
      validator: (value) => propValidator(
        value,
        ['primary', 'secondary'],
      ),
    },
    // Set error styling
    error: {
      type: [Boolean, String],
      default: false,
    },
    disabled: Boolean,
    required: Boolean,
    optional: Boolean,
    value: {
      type: [String, Number],
    },
  },
  data() {
    return {
      style,
      isFocused: false,
    };
  },
  computed: {
    // Use given id or generate one
    inputId() {
      return this.id ? this.id : this._uid; // eslint-disable-line no-underscore-dangle
    },
    baseClass() {
      return 'cdr-input';
    },
    inputAttrs() {
      const isNum = this.type === 'number';
      return {
        pattern: isNum && '[0-9]*',
        inputmode: isNum && 'numeric',
        novalidate: isNum,
        autocorrect: 'off',
        spellcheck: 'false',
        autocapitalize: 'off',
        ...this.$attrs,
      };
    },
    inputClass() {
      const hasPostIcon = !!this.$slots['post-icon'];
      const hasPostIcons = hasPostIcon && this.$slots['post-icon'].length > 1;
      return {
        [this.style['cdr-input']]: true,
        [this.style['cdr-input--multiline']]: this.rows > 1,
        [this.style['cdr-input--preicon']]: this.$slots['pre-icon'],
        [this.style['cdr-input--posticon']]: hasPostIcon,
        [this.style['cdr-input--posticons']]: hasPostIcons,
        [this.style['cdr-input--error']]: this.error,
        [this.style[`cdr-input--${this.background}`]]: true,
      };
    },
    wrapperClass() {
      return {
        [this.style['cdr-input-wrap']]: true,
        [this.style['cdr-input--focus']]: this.isFocused,
      };
    },
    inputListeners() {
      // https://vuejs.org/v2/guide/components-custom-events.html#Binding-Native-Events-to-Components
      // handles conflict between v-model and v-on="$listeners"
      const vm = this;
      return {
        ...this.$listeners,
        input(event) {
          vm.$emit('input', event.target.value);
        },
        focus(event) {
          vm.isFocused = true;
          vm.$emit('focus', event);
        },
        blur(event) {
          vm.isFocused = false;
          vm.$emit('blur', event);
        },
      };
    },
    inputEl() {
      if (this.rows && this.rows > 1) {
        return (
          <textarea
            rows={this.rows}
            class={clsx(
              this.inputClass,
              this.sizeClass,
            )}
            id={this.inputId}
            disabled={this.disabled}
            required={this.required}
            aria-label={this.hideLabel ? this.label : null}
            ref="input"
            {...{ attrs: this.inputAttrs, on: this.inputListeners }}
            vModel={this.value}
          />
        );
      }
      return (
          <input
            type={this.type}
            class={clsx(
              this.inputClass,
              this.sizeClass,
            )}
            id={this.inputId}
            disabled={this.disabled}
            required={this.required}
            aria-label={this.hideLabel ? this.label : null}
            ref="input"
            {...{ attrs: this.inputAttrs, on: this.inputListeners }}
            vModel={this.value}
          />
      );
    },
  },
  render() {
    return (
      <div>
        <cdr-label-standalone
          for-id={ `${this.inputId}` }
          label={ this.label }
          hide-label={ this.hideLabel }
          required={ this.required }
          optional={ this.optional }
          disabled={ this.disabled }
        >
          { this.$slots['helper-text-top'] && (
            <template slot="helper">
              { this.$slots['helper-text-top'] }
            </template>
          )}
          { this.$slots.info && (
            <template slot="info">
              {this.$slots.info}
            </template>
          )}
        </cdr-label-standalone>
        <div class={this.style['cdr-input-outer-wrap']}>
          <div class={this.wrapperClass}>
            {this.inputEl}
            {this.$slots['pre-icon'] && (
              <span
                class={this.style['cdr-input__pre-icon']}
              >
                {this.$slots['pre-icon']}
              </span>
            )}

            {this.$slots['post-icon'] && (
              <span
                class={this.style['cdr-input__post-icon']}
              >
                {this.$slots['post-icon']}
              </span>
            )}
          </div>
          {this.$slots['info-action'] && (
            <div
              class={this.style['cdr-input__info-action']}
            >
              {this.$slots['info-action']}
            </div>
          )}
        </div>
        {(this.$slots['helper-text-bottom'])
          && !this.error && (
            <span class={this.style['cdr-input__helper-text']}>
              {this.$slots['helper-text-bottom']}
            </span>
        )}
        {this.error && (
          <cdr-form-error role={this.errorRole} error={this.error}>
            <template slot="error">
              {this.$slots.error}
            </template>
          </cdr-form-error>
        )}
      </div>
    );
  },
};
