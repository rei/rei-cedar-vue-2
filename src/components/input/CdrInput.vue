<template>
  <div :class="$style['cdr-input-container']">
    <label
      v-if="!hideLabel"
      :class="labelClass"
      :for="inputId"
      ref="label"
    >{{ label }}<span v-if="required">*</span>
    </label>
    <div :class="[inputWrapClass]">
      <textarea
        v-if="multiLine"
        :class="[inputClass, modifierClass]"
        v-bind="$attrs"
        :id="inputId"
        :value="currentValue"
        @blur="onBlur"
        @input="onInput"
        @focus="onFocus"
        @paste="onPaste"
        @keydown="onKeydown"
        :required="required"
        :disabled="disabled"
        :aria-label="hideLabel ? label : null"
        :aria-describedby="errors.length ? messagesId : null"
        ref="input"
      />

      <input
        v-else
        :type="type"
        :class="[inputClass, modifierClass]"
        v-bind="$attrs"
        :id="inputId"
        :value="currentValue"
        @blur="onBlur"
        @input="onInput"
        @focus="onFocus"
        @paste="onPaste"
        @keydown="onKeydown"
        :required="required"
        :disabled="disabled"
        :aria-label="hideLabel ? label : null"
        :aria-describedby="errors.length ? messagesId : null"
        ref="input"
      >
      <span
        v-if="$slots.preicon"
        :class="$style['cdr-input__pre-icon']"
        aria-hidden="true"
      >
        <!-- @slot Icon to be put in front of input -->
        <slot name="preicon"/>
      </span>
      <span
        v-if="feedback"
        :class="validationIconClass"
        v-html="getIcon"
        ref="icon"
      />
    </div>
    <transition-group
      :class="$style['cdr-input-messages']"
      :id="messagesId"
      ref="messages"
      name="cdr-animated-errors"
      tag="div"
    >
      <div
        :class="messageClass"
        ref="error"
        v-for="error in errors"
        :error="error"
        :key="error">{{ error }}</div>
    </transition-group>
  </div>
</template>

<script>
import debounce from 'srcdir/utils/debounce';
import modifier from 'mixinsdir/modifier';

const checkIcon = require('assetsdir/icons/rei/check-lg.svg');
const errorIcon = require('assetsdir/icons/rei/x-fill.svg');
const warningIcon = require('assetsdir/icons/rei/warning-tri.svg');

/**
 * Cedar 2 component for input
 * **NOTE:** `v-model` is required.
 * @version 0.0.1
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrInput',
  mixins: [modifier],
  inheritAttrs: false,
  props: {
    /**
     * `id` for the input that is mapped to the label `for` attribute. If one is not provided, it will be auto generated.
    */
    id: String,
    /**
     * Label text. This is required for a11y even if hiding the label with `hideLabel`.
    */
    label: {
      type: String,
      required: true,
    },
    /**
     * Removes the label element but sets the input `aria-label` to `label` text for a11y.
    */
    hideLabel: Boolean,
    /**
     * Changes the input to a textarea.
    */
    multiLine: Boolean,
    /**
     * Regex validation pattern. Useful for simple validation.
    */
    pattern: String,
    /**
     * Error message to be displayed when `pattern` validation fails.
    */
    patternError: String,
    /**
     * Enables icon feedback as part of validation for valid, warn, and error states.
    */
    feedback: Boolean,
    /**
     * Input type. NOTE: This component is meant for text style inputs. Other input types (checkbox, radio) have their own components.
    */
    type: {
      type: String,
      default: 'text',
    },
    /**
     * Array of functions. Provide your own validation function(s). Takes the input string and outputs an object with a state and message. State is a String with value `valid`, `warning`, or `error`. Message is any String.
    */
    rules: {
      type: Array,
      default: () => [],
    },
    /**
     * Boolean or Number. `true` has a default of 500ms. Providing a number will set debounce to that (in ms). `false` is no debounce.
    */
    debounce: {
      type: [Boolean, Number],
      required: false,
      default: false,
    },
    /** @ignore */
    disabled: Boolean,
    /** @ignore */
    required: Boolean,
    /** @ignore */
    immediateValidate: Boolean,
    /** @ignore */
    value: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
      required: false,
    },
  },
  data() {
    return {
      errors: [],
      currentValue: this.value,
      pristine: true,
      touched: false,
      valid: false,
      focused: false,
      state: '',
    };
  },
  computed: {
    // Use given id or generate one
    inputId() {
      return this.id ? this.id : this._uid; // eslint-disable-line no-underscore-dangle
    },
    messagesId() {
      return `err${this._uid}`; // eslint-disable-line no-underscore-dangle
    },
    baseClass() {
      return 'cdr-input';
    },
    labelClass() {
      return {
        [this.$style['cdr-input__label']]: true,
        [this.$style['cdr-input__label--error']]: this.isErr,
        [this.$style['cdr-input__label--disabled']]: this.disabled,
      };
    },
    inputClass() {
      return {
        [this.$style['cdr-input']]: true,
        [this.$style['cdr-input--multiline']]: this.multiLine,
        [this.$style['cdr-input--error']]: this.isErr,
        [this.$style['cdr-input--warn']]: this.isWarn,
        [this.$style['cdr-input--preicon']]: this.$slots.preicon,
      };
    },
    inputWrapClass() {
      return {
        [this.$style['cdr-input-wrap']]: true,
        [this.$style['cdr-input-wrap--actions']]: this.$slots.pre || this.$slots.post,
      };
    },
    validationIconClass() {
    /* istanbul ignore next */
      return {
        [this.$style['cdr-input-wrap__icon']]: true,
        [this.$style['cdr-input-wrap__icon--error']]: this.isErr,
        [this.$style['cdr-input-wrap__icon--warn']]: this.isWarn,
        [this.$style['cdr-input-wrap__icon--valid']]: this.isValid,
      };
    },
    messageClass() {
      return {
        [this.$style['cdr-input-messages__notification']]: true,
        [this.$style['cdr-input-messages__notification--error']]: this.isErr,
        [this.$style['cdr-input-messages__notification--warn']]: this.isWarn,
      };
    },
    getIcon() {
      let icon = '';

      if (this.isErr) {
        icon = errorIcon;
      } else if (this.isValid) {
        icon = checkIcon;
      } else if (this.isWarn) {
        icon = warningIcon;
      }

      return icon;
    },
    // Check if debounce is enabled, defined, or default
    debounceVal() {
      if (this.debounce === false) {
        return 0;
      } else if (!Number.isNaN(this.debounce) && this.debounce !== '' && this.debounce !== true) {
        return this.debounce;
      }
      return 500;
    },
    isErr() {
      return this.state === 'error';
    },
    isWarn() {
      return this.state === 'warn';
    },
    isValid() {
      return this.state === 'valid';
    },
  },
  watch: {
    focused(val) {
      this.touched = true;

      /**
      * New input value (if changed). Fires on blur.
      * @event change
      * @type string
      */
      if (val) {
        this.$emit('change', val);
      }
    },
    value(val) {
      this.setCurrentValue(val);
      this.validate();
    },
  },
  mounted() {
    // Convert pattern to a rule for testing
    if (this.pattern) {
      const regPattern = new RegExp(this.pattern);
      this.rules.push((text) => {
        const obj = {};
        if (regPattern.test(text)) {
          obj.state = 'valid';
        } else {
          obj.state = 'error';
          obj.message = this.patternError || '';
        }
        return obj;
      });
    }
    // Provide some default validation for required
    if (this.required) {
      this.rules.push((text) => {
        const obj = {};
        // interacted with, not currently focsed, and empty
        if (this.touched && !this.focused && text === '') {
          obj.state = 'warn';
          obj.message = 'This field is required';
        }
        return obj;
      });
    }

    if (this.immediateValidate) {
      this.validate(true);
    }
  },
  methods: {
    onInput(e) {
      this.pristine = false;
      const { value } = e.target;
      /**
      * Current input value. Fires while typing.
      * Returns (value, event)
      * @event input
      * @type {event}
      */
      this.$emit('input', value, e);
      this.setCurrentValue(value);
    },
    onBlur(e) {
      this.validate(true);
      this.$nextTick(() => {
        this.focused = false;
      });
      /**
      * Fires when input loses focus.
      * @event blur
      * @type {event}
      */
      this.$emit('blur', e);
    },
    onFocus(e) {
      /**
      * Fires when input gains focus.
      * @event focus
      * @type {event} */
      this.focused = true;
      this.$emit('focus', e);
    },
    onPaste(e) {
      /**
      * Fires when text is pasted into input.
      * @event paste
      * @type {event}
       */
      this.validate(true);
      this.$emit('paste', e);
    },
    onKeydown(e) {
      /**
      * Fires when a key is pressed.
      * @event keydown
      * @type {event}
       */
      this.$emit('keydown', e);
    },
    setCurrentValue(value) {
      if (value === this.currentValue) return;
      this.currentValue = value;
    },
    validate(immediate = false) {
      // only validate rules if there are any
      if (this.rules.length > 0) {
        const delay = immediate ? 0 : this.debounceVal;
        (debounce(() => {
          this.errors = [];
          this.valid = false;

          this.rules.forEach((rule) => {
            const validObj = rule(this.currentValue);
            validObj.state = validObj.state ? validObj.state : this.state;

            if (validObj.state === 'valid') {
              this.state = validObj.state;
            } else if (validObj.state === 'warn') {
              this.state = validObj.state;
              this.errors.push(validObj.message);
            } else {
              this.state = validObj.state;
              if (validObj.message) {
                this.errors.push(validObj.message);
              }
            }
          });
        }, delay))();
      }
    },
  },
};
</script>

<style module>
  @import 'cssdir/settings/_index.pcss';
  @import './styles/vars/CdrInput.vars.pcss';
  @import './styles/CdrInput.pcss';
</style>
