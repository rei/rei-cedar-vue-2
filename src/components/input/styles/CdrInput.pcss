/* ==========================================================================
  # CdrInput

  All values should map to variables in
  vars/CdrInput.vars.pcss
  in order to allow for theming

  TOC:

    :Base - Label
    :Base - Input container
    :Base - Input
      :States
      :Modifiers
        :Style Variants
    :Base - Media Frame
      :Modifiers
        :Aspect Ratios
          :normal
          :@small
          :@medium
          :@large
        :Cropping
      :Elements
========================================================================== */

/* ==========================================================================
  # INPUT LABEL
  ========================================================================== */
.cdr-input__label {
  @include maple-utility-20;
  
  color: $input-label-color-text;
  margin-bottom: $input-label-margin-bottom;

  &--error {
    color: $input-label-error-color-text;
  }

  &--disabled {
    color: $input-label-disabled-color-text;
  }
}

/* ==========================================================================
  # INPUT CONTAINER
  ========================================================================== */

.cdr-input-container {
  & + & {
    margin-top: $input-container-spacing;
  }
}

/* ==========================================================================
  # INPUT
  ========================================================================== */

.cdr-input {
  font-size: $input-font-size;
  line-height: $input-line-height;
  text-align: $input-text-align;
  color: $input-color-text;
  font-style: $input-font-style;
  font-weight: $input-font-weight;
  font-family: $input-font-family;
  background-color: $input-background-color;
  border-color: $input-border-color;
  border-style: $input-border-style;
  border-radius: $input-border-radius;
  border-width: $input-border-width;
  padding: $input-padding;
  height: $input-height;
  display: block;
  width: 100%;

  /* States
  ---------- */

  &[disabled] {
    background-color: $input-disabled-background-color;
    color: $input-disabled-text-color;
  }

  /* Style variants
    ========================================================================== */

  &--multiline {
    height: auto;
  }

  &--error {
    border-color: $input-error-border-color;
    border-bottom-right-radius: 0;
    border-bottom-left-radius: 0;
  }

  &--warn {
    border-color: $input-warning-border-color;
    border-bottom-right-radius: 0;
    border-bottom-left-radius: 0;
  }

  &--preicon {
    padding-left: $input-preicon;
  }

  /* Size variants
    ========================================================================== */

  &--large {
    padding-top: $inset-md;
    padding-bottom: $inset-md;
    height: auto;
  }

  /* Pseudo elements
    ========================================================================== */

  &::placeholder {
    font-style: italic;
    font-size: $input-size-placeholder-text;
    line-height: $input-line-height-placeholder;
    color: $input-color-placeholder-text;
  }

  /* Block elements
    ========================================================================== */

  &__pre-icon {
    position: absolute;
    top: 50%;
    left: $sizing-size-md;
    transform: translateY(-50%);
  }
}

/* ==========================================================================
   # INPUT WRAPPER
   ========================================================================== */

/*
* TOC:
*
* Base class
* Style variants
* Block elements
*/

.cdr-input-wrap {
  position: relative;

  /* Block elements
    ========================================================================== */

  &__icon {
    position: absolute;
    top: $input-icon-top;
    right: $input-icon-right;
    height: $input-icon-height;
    width: $input-icon-width;

    &--valid > svg {
      fill: $input-icon-success-fill;
    }

    &--warn > svg {
      fill: $input-warning-border-color;
    }

    &--error > svg {
      fill: $input-error-border-color;
    }
  }
}

/* ==========================================================================
   # ANIMATED ERRORS
   ========================================================================== */

.cdr-animated-errors {
  &-enter-active,
  &-leave-active {
    transition: opacity 350ms ease, transform 350ms ease;
    transform-origin: top;
  }

  &-leave-active + &-enter-active {
    transition-delay: 350ms;
    position: absolute;
    width: 100%;
  }

  &-enter,
  &-leave-to {
    opacity: 0;
    transform: scaleY(0);
  }
}

/* ==========================================================================
   # INPUT MESSAGES
   ========================================================================== */

/*
* TOC:
*
* Base class
* Block elements
*/

.cdr-input-messages {
  position: relative;
  font-size: $sys-font-redwood-ui-20-font-size;
  line-height: $sys-font-redwood-ui-20-line-height;
  text-align: $sys-font-redwood-ui-20-text-align;
  color: $sys-font-redwood-ui-20-color;
  font-style: $sys-font-redwood-ui-20-font-style;
  font-weight: $sys-font-redwood-ui-20-font-weight;
  font-family: $sys-font-redwood-ui-20-font-family;

  /* Block elements
    ========================================================================== */

  &__notification {
    padding: $inset-sm;
    border-style: solid;
    border-width: 1px;
    border-top: none;

    &:last-of-type {
      border-bottom-right-radius: $radius-input;
      border-bottom-left-radius: $radius-input;
    }

    &--error {
      border-color: $input-error-border-color;
      background-color: $input-error-background-color;
    }

    &--warn {
      border-color: $input-warning-border-color;
      background-color: $input-warning-background-color;
    }
  }
}
