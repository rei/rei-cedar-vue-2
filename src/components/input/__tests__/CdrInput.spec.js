import { shallowMount, mount } from '@vue/test-utils';
import sinon from 'sinon';
import CdrInput from 'componentdir/input/CdrInput';

describe('CdrInput', () => {
  it('renders a label element', () => {
    const wrapper = mount(CdrInput, {
      propsData: {
        label: 'Label Test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders required label correctly', () => {
    const wrapper = mount(CdrInput, {
      propsData: {
        label: 'Label Test',
        required: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('maps input id to label for correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'testing',
        id: 'test',
      },
    });
    expect(wrapper.vm.$refs.input.id).toBe(wrapper.vm.$refs.label.htmlFor);
  });

  it('generates an id correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'testing',
      },
    });
    expect(wrapper.vm.$refs.input.id).toBe(wrapper.vm._uid.toString());
  });

  it('renders an input element', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'Label Test',
      },
    });
    expect(wrapper.vm.$refs.input.tagName).toBe('INPUT');
  });

  it('sets input name attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'Label Test',
      },
      attrs: {
        name: 'testName',
      }
    });
    expect(wrapper.vm.$refs.input.hasAttribute('name', 'testName')).toBe(true);
  });

  it('sets input disabled attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
        disabled: true,
      },
    });
    expect(wrapper.vm.$refs.input.hasAttribute('disabled')).toBe(true);
  });

  it('sets input readonly attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
      attrs: {
        readonly: true,
      },
    });
    expect(wrapper.vm.$refs.input.hasAttribute('readonly')).toBe(true);
  });

  it('sets input required attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
        required: true,
      },
    });
    expect(wrapper.vm.$refs.input.hasAttribute('required')).toBe(true);
  });

  it('sets input autofocus attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
      attrs: {
        autofocus: true,
      },
    });
    expect(wrapper.vm.$refs.input.hasAttribute('autofocus')).toBe(true);
  });

  it('sets input maxlength attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
      attrs: {
        maxlength: '20',
      },
    });
    expect(wrapper.vm.$refs.input.hasAttribute('maxlength', '20')).toBe(true);
  });

  it('sets input placeholder correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
      attrs: {
        placeholder: 'test placeholder',
      },
    });
    expect(wrapper.vm.$refs.input.hasAttribute('placeholder', 'test placeholder')).toBe(true);
  });

  it('sets multiline to textarea correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
        rows: 2,
      },
    });
    expect(wrapper.vm.$refs.input.tagName).toBe('TEXTAREA');
  });

  it('sets multiline rows correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
        multiLine: true,
        rows: 10
      },
    });
    expect(wrapper.vm.$refs.input.hasAttribute('rows', '10')).toBe(true);
  });

  it('sets default input type correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
    });
    expect(wrapper.vm.$refs.input.hasAttribute('type', 'text')).toBe(true);
  });

  it('overrides input type correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
        type: 'url',
      },
    });
    expect(wrapper.vm.$refs.input.hasAttribute('type', 'url')).toBe(true);
  });

  it('hide-label sets aria-label correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
        hideLabel: true,
      },
    });
    expect(wrapper.vm.$refs.input.hasAttribute('aria-label', 'test')).toBe(true);
  });

  it('auto generates an id', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
    });
    expect(wrapper.vm.inputId).toBe(wrapper.vm._uid); // eslint-disable-line no-underscore-dangle
  });

  it('emits an input event', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
        value: 'bar'
      },
    });
    const input = wrapper.find({ ref: 'input' });
    input.setValue('foo');
    expect(wrapper.emitted().input[0][0]).toBe('foo');
  });

  it('emits a blur event', () => {
    const spy = sinon.spy();
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
      listeners: {
        'blur': spy
      }
    });
    const input = wrapper.find({ ref: 'input' });
    input.trigger('blur')
    expect(spy.calledOnce).toBeTruthy();
  });

  it('emits a focus event', () => {
    const spy = sinon.spy();
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
      listeners: {
        'focus': spy
      }
    });
    const input = wrapper.find({ ref: 'input' });
    input.trigger('focus')
    expect(spy.calledOnce).toBeTruthy();
  });

  it('emits a paste event', () => {
    const spy = sinon.spy();
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
      listeners: {
        'paste': spy
      }
    });
    const input = wrapper.find({ ref: 'input' });
    input.trigger('paste')
    expect(spy.calledOnce).toBeTruthy();
  });

  it('emits a keydown event', () => {
    const spy = sinon.spy();
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test'
      },
      listeners: {
        'keydown': spy
      }
    });
    const input = wrapper.find({ ref: 'input' });
    input.trigger('keydown', {
      key: 'a'
    })
    expect(spy.called).toBeTruthy();
  });

  it('renders helper-text slot', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
      slots: {
        'helper-text': 'very helpful',
      },
    });
    expect(wrapper.find('.cdr-input__helper-text').text()).toBe('very helpful');
  });

  it('renders info slot', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
      slots: {
        info: 'very informative',
      },
    });
    expect(wrapper.find('.cdr-input__info-container').text()).toBe('very informative');
  });

  it('renders pre-icon slot', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
      slots: {
        'pre-icon': 'ðŸ¤ ',
      },
    });
    expect(wrapper.find('.cdr-input__pre-icon').text()).toBe('ðŸ¤ ');
  });

  it('renders post-icon slot', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
      },
      slots: {
        'post-icon': 'ðŸ˜Ž',
      },
    });
    expect(wrapper.find('.cdr-input__post-icon').text()).toBe('ðŸ˜Ž');
  });

  // NOTE - can't use v-model directly here, targeting the `data` prop instead
  it('updating v-model data updates the input', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
        value: 'bar'
      },
    });
    const input = wrapper.find({ ref: 'input' });
    wrapper.setProps({value: ''});
    expect(input.element.value).toBe('');
  });
});
