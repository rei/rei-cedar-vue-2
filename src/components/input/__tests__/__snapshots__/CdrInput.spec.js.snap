// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrInput renders a label element 1`] = `
<div>
  <label
    class="cdr-input__label"
    for="1"
  >
    Label Test
  </label>
  <div
    class="cdr-input-wrap"
  >
    <input
      class="cdr-input cdr-input "
      id="1"
      type="text"
    />
  </div>
</div>
`;

exports[`CdrInput renders required label correctly 1`] = `
<div>
  <label
    class="cdr-input__label"
    for="3"
  >
    Label Test
    <span
      class="cdr-input__required-label"
    >
      Required
    </span>
  </label>
  <div
    class="cdr-input-wrap"
  >
    <input
      class="cdr-input cdr-input "
      id="3"
      required="required"
      type="text"
    />
  </div>
</div>
`;
